src/connector/src/sink/mongodb.rs

```diff
@@ -0,0 +1,860 @@
+        } else {
+            self.pk_indices
+                .iter()
+                .map(|&idx| {
+                    let pk_field = &self.schema.fields[idx];
+                    (
+                        pk_field.name.clone(),
+                        bson_from_scalar_ref(pk_field, row.datum_at(idx)),
+                    )
+                })
+                .collect::<Document>()
+                .into()
+        }
+    }
+
+    fn append<'a>(&'a mut self, row: RowRef<'a>) {
+        let document = self.document_from_row_ref(row);
+        let ns = self.extract_namespace_from_row_ref(row);
+        let coll = ns.1.clone();
+        match self.insert_builder.as_mut().unwrap().entry(ns) {
+            Entry::Occupied(mut entry) => entry.get_mut().append(document),
+            Entry::Vacant(entry) => {
+                let mut builder = InsertCommandBuilder::new(coll, self.max_entries);
+                builder.append(document);
+                entry.insert(builder);
+            }
+        }
+    }
+
+    fn upsert<'a>(&'a mut self, op: Op, row: RowRef<'a>) -> Result<()> {
+        let mut document = self.document_from_row_ref(row);
+        let ns = self.extract_namespace_from_row_ref(row);
+        let coll = ns.1.clone();
+
+        let pk = self.construct_pk(row);
+
+        // Specify the primary key (_id) for the MongoDB collection if the user does not provide one.
+        if self.pk_indices.len() > 1
+            || self.schema.fields[self.pk_indices[0]].name != MONGODB_PK_NAME
+        {
+            // compound pk should not have a field named `_id`
+            document.insert(MONGODB_PK_NAME, pk.clone());
+        }
+
+        let pk = doc! {MONGODB_PK_NAME: pk};
+        match op {
+            Op::Insert | Op::UpdateInsert => {
+                match self.upsert_builder.as_mut().unwrap().entry(ns) {
+                    Entry::Occupied(mut entry) => entry.get_mut().add_upsert(pk, document),
+                    Entry::Vacant(entry) => {
+                        let mut builder = UpsertCommandBuilder::new(coll, self.max_entries);
+                        builder.add_upsert(pk, document)?;
+                        entry.insert(builder);
+                        Ok(())
+                    }
+                }
+            }
+            Op::UpdateDelete => Ok(()),
+            Op::Delete => match self.upsert_builder.as_mut().unwrap().entry(ns) {
+                Entry::Occupied(mut entry) => entry.get_mut().add_delete(pk),
+                Entry::Vacant(entry) => {
+                    let mut builder = UpsertCommandBuilder::new(coll, self.max_entries);
+                    builder.add_delete(pk)?;
+                    entry.insert(builder);
+                    Ok(())
+                }
+            },
+        }
+    }
+
+    async fn write<'a>(&'a mut self, op: Op, row: RowRef<'a>) -> Result<()> {
+        if self.is_append_only {
+            if op != Op::Insert {
+                return Ok(());
```

Comment:
we should panic for this case, which is a bug.

Reply:
Agree, actually the idea is borrowed from mqtt sink and starrocks sink, they all skipped the `Op::Insert`. Should we panic here?

Reply:
Yeah, just panic may be too rude although it could be a bug. For the sake of robustness, I think we can skip the INSERT op but also print a WARN log.

Reply:
WARN log is now printed when a non-insert op received in append-only mode.